Support 'number' as argument for filter $type

guludo opened the pull request:
Hi there.

This is a simple PR adding support for `number` as argument for filter operator `$type`, which is an alias grouping numeric types. 

For reference: https://docs.mongodb.com/manual/reference/operator/query/type/. It looks like this is supported by MongoDB from at least version 4.0 (haven't checked documentation for older versions).

Regards,  
Gustavo Sousa

guludo committed:
Support 'number' as argument for filter $type

pcorpet reviewed (changes requested):
Very cool, thanks.

Please add a test also in test__mongomock to check that the behavior is the same with pymongo

guludo committed:
Fix behavior for $type filter

- The `$type` filter also matches  against array items;
- Python consider a boolean value to be instance of int, so we need to
  catch that for `int` and `long`
- Check `bit_length()` of Python integers in order to evaluate for `int`
  and `long`
- Use proper type for `decimal`
- Use proper aliases for `number`

guludo commented:
Hi @pcorpet !

As I was working in test__mongomock to add tests for `$type`, the tests showed that the current implementation already failed to reproduce mongo for other types, so I went ahead and added a commit to fix the behavior for them.

pcorpet commented:
thanks a lot for finding those. Would you be ok to split the fix in a different PR please? and have this PR only to introduce `number`. I like having a clean history and have only one commit per PR.

pcorpet commented on tests/test__mongomock.py:
```diff
@@ -1013,6 +1013,26 @@ def test__find_projection_subdocument_lists(self):
+            {'a': True},  # bool
+            {'a': datetime.datetime.now()},  # date
+        ])
+        for type_name, spec in mongomock.filtering.TYPE_MAP.items():
```
I'd rather not rely on mongomock internals. Please list all the supported types here in a list.

guludo replied:
Okay.

Btw, is that the same for `test__empty_logical_operators`?

pcorpet replied:
You're right, I'll fix it.

guludo commented:
> thanks a lot for finding those. Would you be ok to split the fix in a different PR please?

Ok.

guludo commented:
I'm closing this one and will open separate pull requests.
